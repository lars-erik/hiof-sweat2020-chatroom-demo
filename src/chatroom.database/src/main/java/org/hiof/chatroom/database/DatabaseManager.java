package org.hiof.chatroom.database;

import java.sql.*;

public class DatabaseManager {
    public static void ensureDatabase(String path) {
        String url = "jdbc:sqlite:" + path;

        try {
            Class.forName("org.sqlite.JDBC");
            Connection conn = DriverManager.getConnection(url);
            if (conn != null) {
                DatabaseMetaData meta = conn.getMetaData();
            }

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }
}
