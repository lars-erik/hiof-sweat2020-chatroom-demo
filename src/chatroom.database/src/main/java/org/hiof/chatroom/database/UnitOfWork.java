package org.hiof.chatroom.database;

import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.Transaction;
import org.hibernate.boot.MetadataBuilder;
import org.hibernate.boot.MetadataSources;
import org.hibernate.boot.registry.StandardServiceRegistry;
import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
import org.hibernate.type.UUIDCharType;

import java.util.UUID;

public class UnitOfWork implements org.hiof.chatroom.persistence.UnitOfWork {
    public Session getSession() {
        return session;
    }

    public UnitOfWork() throws Exception {
        this.session = getSessionFactory().openSession();
        this.transaction = this.session.beginTransaction();
    }

    @Override
    public void saveChanges() {
        getSession().flush();
        this.transaction.commit();
        this.transaction = this.session.beginTransaction();
    }

    @Override
    public void close() {
        getSession().close();
    }

    private final Session session;
    private Transaction transaction;

    static SessionFactory sessionFactory;
    static SessionFactory getSessionFactory() throws Exception {
        if (sessionFactory != null) {
            return sessionFactory;
        }

        // A SessionFactory is set up once for an application!
        final StandardServiceRegistry registry = new StandardServiceRegistryBuilder()
                .configure() // configures settings from hibernate.cfg.xml
                .build();
        try {
            sessionFactory = new MetadataSources( registry ).buildMetadata().buildSessionFactory();
        }
        catch (Exception e) {
            // The registry would be destroyed by the SessionFactory, but we had trouble building the SessionFactory
            // so destroy it manually.
            StandardServiceRegistryBuilder.destroy( registry );
            throw e;
        }

        return sessionFactory;
    }
}
