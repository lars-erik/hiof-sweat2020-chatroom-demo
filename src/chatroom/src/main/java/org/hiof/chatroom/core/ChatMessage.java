package org.hiof.chatroom.core;

import java.time.*;
import java.time.format.*;
import java.util.UUID;

public class ChatMessage {
    private UUID id;
    private Instant time;
    private String user;
    private String message;

    public ChatMessage() {
    }

    public ChatMessage(UUID id, Instant time, String user, String message) {
        this.id = id;
        this.time = time;
        this.user = user;
        this.message = message;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public Instant getTime() {
        return time;
    }

    public void setTime(Instant time) {
        this.time = time;
    }

    @Override
    public String toString() {
        return "ChatMessage{" +
                "time='" + time + '\'' +
                ", user='" + user + '\'' +
                ", message='" + message + '\'' +
                '}';
    }

    DateTimeFormatter formatter = DateTimeFormatter
            .ofPattern("HH:mm:ss")
            .withZone(ZoneId.of("UTC"));

    public String format() {
        String timeString = formatter.toFormat().format(getTime());
        return String.format("%1$s <%2$s>: %3$s", timeString, getUser(), getMessage());
    }
}
