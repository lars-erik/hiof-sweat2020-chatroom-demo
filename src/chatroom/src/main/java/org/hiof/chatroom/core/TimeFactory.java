package org.hiof.chatroom.core;

import java.time.Instant;
import java.util.concurrent.Callable;

public class TimeFactory {
    public static Callable<Instant> nowFactory = () -> Instant.now();
    public static Instant now() throws Exception { return nowFactory.call(); }
    public static void reset() {
        nowFactory = () -> Instant.now();
    }
}
