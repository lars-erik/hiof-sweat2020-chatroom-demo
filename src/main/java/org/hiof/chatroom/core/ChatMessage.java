package org.hiof.chatroom.core;

import java.util.UUID;

public class ChatMessage {
    private UUID id;
    private String user;
    private String message;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }
}
