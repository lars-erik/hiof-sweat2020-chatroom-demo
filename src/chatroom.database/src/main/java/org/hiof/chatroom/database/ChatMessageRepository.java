package org.hiof.chatroom.database;

import org.hibernate.Query;
import org.hibernate.Session;
import org.hiof.chatroom.core.ChatMessage;

import java.util.List;
import java.util.UUID;
import java.util.stream.Stream;

public class ChatMessageRepository implements org.hiof.chatroom.persistence.ChatMessageRepository {

    private final UnitOfWork unitOfWork;

    public ChatMessageRepository(UnitOfWork unitOfWork) {
        this.unitOfWork = unitOfWork;
    }

    @Override
    public void add(ChatMessage message) {
        unitOfWork.getSession().save(message);
    }

    @Override
    public Stream<ChatMessage> query() {
        Session session = unitOfWork.getSession();
        Query query = session.createQuery("SELECT msg FROM ChatMessage msg");
        List<ChatMessage> msgs = query.list();
        return msgs.stream();
    }

    @Override
    public ChatMessage get(String id) {
        return unitOfWork.getSession().get(ChatMessage.class, id);
    }
}
