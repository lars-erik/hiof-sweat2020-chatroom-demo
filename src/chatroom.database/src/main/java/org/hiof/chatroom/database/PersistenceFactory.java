package org.hiof.chatroom.database;

import org.hiof.chatroom.persistence.ChatMessageRepository;
import org.hiof.chatroom.persistence.UnitOfWork;

public class PersistenceFactory extends org.hiof.chatroom.persistence.PersistenceFactory {
    @Override
    public UnitOfWork createUnitOfWork() throws Exception {
        return new org.hiof.chatroom.database.UnitOfWork();
    }

    @Override
    public ChatMessageRepository createChatMessageRepository(UnitOfWork uow) {
        return new org.hiof.chatroom.database.ChatMessageRepository((org.hiof.chatroom.database.UnitOfWork)uow);
    }
}
