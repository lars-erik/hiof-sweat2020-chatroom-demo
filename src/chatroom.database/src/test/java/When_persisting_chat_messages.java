import org.approvaltests.Approvals;
import org.hiof.chatroom.core.ChatMessage;
import org.hiof.chatroom.database.*;
import org.junit.jupiter.api.*;

import java.time.Instant;
import java.util.UUID;

public class When_persisting_chat_messages {
    @Test
    public void stores_message() throws Exception {
        DatabaseManager.ensureDatabase("./db/chat-test.db");
        UnitOfWork uow = new UnitOfWork();
        ChatMessageRepository repo = new ChatMessageRepository(uow);
        UUID id = UUID.fromString("70f49e80-bf70-44bb-8179-b3f2a1357a11");
        ChatMessage msg = new ChatMessage(
            id, Instant.parse("2022-10-15T12:00:00Z"),
            "Luke", "This is red leader..."
        );
        repo.add(msg);
        uow.saveChanges();
        uow.close();
        uow = new UnitOfWork();
        repo = new ChatMessageRepository(uow);
        msg = repo.get(id);
        Approvals.verify(msg);
    }
}
