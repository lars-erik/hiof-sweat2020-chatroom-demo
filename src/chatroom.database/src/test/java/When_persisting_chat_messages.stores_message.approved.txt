ChatMessage{time='2022-10-15T12:00:00Z', user='Luke', message='This is red leader...'}